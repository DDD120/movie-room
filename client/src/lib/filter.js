import dayjs from "dayjs";

export const arrayDeduplication = (list) => {
  return list.filter(
    (item, index, array) =>
      index === array.findIndex((data) => data.id === item.id)
  );
};

export const getYear = (date) => {
  return dayjs(date).year();
};

export const isNull = (string) => {
  const blank_pattern = /^\s+|\s+$/g;
  return string.replace(blank_pattern, "") === "";
};
