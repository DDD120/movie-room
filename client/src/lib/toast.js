import { toast } from "react-toastify";

export const showToast = (message) => {
  toast(message, {
    autoClose: 3000,
    hideProgressBar: true,
  });
};
