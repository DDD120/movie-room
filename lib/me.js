export const me = (user) => {
  return {
    id: user._id,
    profile: {
      nickname: user.profile.nickname,
      thumbnail: user.profile.thumbnail,
    },
    email: user.email,
    reviews: user.reviews,
  };
};
