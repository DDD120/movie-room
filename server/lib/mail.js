import nodemailer from "nodemailer";
import dotenv from "dotenv";

dotenv.config();

export const mailSender = async (email, certificationNumber) => {
  let transporter = nodemailer.createTransport({
    service: "gmail",
    host: "smtp.gmail.com",
    port: 587,
    secure: false,
    auth: {
      user: process.env.NODEMAILER_USER,
      pass: process.env.NODEMAILER_PASS,
    },
  });

  try {
    let info = await transporter.sendMail({
      from: `"MOVIE ROOM" <${process.env.NODEMAILER_USER}>`,
      to: email,
      subject: "[MOVIE ROOM] 회원가입 인증번호 발송",
      html: `
        <body>
          <h1>[MOVIE ROOM]</h1>
          </br>
          <p>MOVIE ROOM에서 인증번호를 발송해 드립니다.</p>
          <div style="padding: 12px 0;">
            <span style="font-size: 24px; font-weight; bold;">${certificationNumber}</span>
          </div>
        </body>
      `,
    });

    return info.accepted;
  } catch (e) {
    throw new Error("메일 전송에 실패하였습니다.", console.error(e));
  }
};
