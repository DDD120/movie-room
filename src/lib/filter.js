export const arrayDeduplication = (list) => {
  return list.filter(
    (item, index, array) =>
      index === array.findIndex((data) => data.id === item.id)
  );
};
