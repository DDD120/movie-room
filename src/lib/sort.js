import dayjs from "dayjs";

export const sortArray = (array, mode) => {
  return sort[mode](array);
};

const sort = {
  newest: (array) =>
    array.sort((a, b) => dayjs(b.updatedAt) - dayjs(a.updatedAt)),
  oldest: (array) =>
    array.sort((a, b) => dayjs(a.updatedAt) - dayjs(b.updatedAt)),
  starDesc: (array) => array.sort((a, b) => b.rating - a.rating),
  starAsc: (array) => array.sort((a, b) => a.rating - b.rating),
};
