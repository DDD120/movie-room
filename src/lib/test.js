import React from "react";
import { render } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import userReducer from "store/user";
import { BrowserRouter } from "react-router-dom";
import { serverApi } from "apis/server-api";
import { moviedbApi } from "apis/movie-db-api";

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    store = configureStore({
      reducer: {
        user: userReducer,
        [serverApi.reducerPath]: serverApi.reducer,
        [moviedbApi.reducerPath]: moviedbApi.reducer,
      },
      preloadedState,
    }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <BrowserRouter>{children}</BrowserRouter>
      </Provider>
    );
  }

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
