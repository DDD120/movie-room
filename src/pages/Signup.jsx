import styled from "@emotion/styled";
import { useEmailMutation } from "apis/server-api";
import Container from "components/common/Container";
import SuccessSendMail from "components/modal/SuccessSnedMail/SuccessSendMail";
import { useState, useEffect, useCallback } from "react";
import { useForm } from "react-hook-form";
import { Link } from "react-router-dom";
import { Common } from "styles/common";
import { toast } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";

const Layout = styled.div`
  display: flex;
  justify-content: center;
  flex-direction: column;
  align-items: center;
`;

const Logo = styled.div`
  width: 200px;
  margin: 40px 0 10px;
  & > img {
    width: 100%;
  }
`;

const Head = styled.h1`
  font-weight: 700;
  font-size: 1.5rem;
  color: ${Common.colors.black};
  margin: 10px 0;
`;

const Form = styled.form`
  max-width: 600px;
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
`;

const Input = styled.input`
  margin: 4px auto;
  width: 100%;
  border-radius: 50px;
  padding: 14px 30px;
  border: none;
  box-shadow: none;
  color: #fff;
  font-size: 1.025rem;
  cursor: ${({ submit }) => submit && "pointer"};
  background-color: ${({ submit }) =>
    submit ? `${Common.colors.cyan}` : `${Common.colors.orange}`};
  &::placeholder {
    color: #fff;
  }
  &:-webkit-autofill,
  :-webkit-autofill:hover,
  :-webkit-autofill:focus,
  :-webkit-autofill:active {
    box-shadow: 0 0 0 1000px ${Common.colors.orange} inset;
    transition: background-color 5000s ease-in-out 0s;
    -webkit-box-shadow: 0 0 0 1000px ${Common.colors.orange} inset;
    -webkit-text-fill-color: #fff;
    -webkit-transition: background-color 5000s ease-in-out 0s;
  }
`;

const ToLogin = styled.span`
  display: block;
  margin-bottom: 10px;
`;

const ErrorMsg = styled.p`
  color: #da0000;
  text-align: left;
  padding: 0 30px;
`;

const EMAIL_REGEX = /^[a-zA-Z0-9]+@+[a-zA-Z0-9]+.+[A-z]/;

const PW_REGEX = /^(?=.*\d)(?=.*[a-zA-Z])[0-9a-zA-Z]{6,10}$/;

const ERROR_MSG = {
  required: "필수정보입니다.",
  invalidEmail: "이메일 형식이 아닙니다.",
  invalidPw: "6~10자 영문, 숫자 조합으로 입력하세요.",
  invalidConfirmPw: "비밀번호와 일치하지 않습니다.",
};

const Signup = () => {
  const [isOpenModal, setIsOpenModal] = useState(false);
  const [emailTrigger, { data: emailRes = {} }] = useEmailMutation();

  const {
    register,
    handleSubmit,
    getValues,
    trigger,
    formState: { errors, isSubmitting },
  } = useForm({ mode: "onChange" });

  const onSubmit = async (data) => {
    emailTrigger({ email: data.email });
  };

  const closeHandler = () => {
    setIsOpenModal(false);
  };

  const showErrorNotify = useCallback((msg) => {
    toast.error(msg, {
      autoClose: 3000,
      hideProgressBar: true,
    });
  }, []);

  useEffect(() => {
    emailRes.type === "SUCCESS_SEND_EMAIL"
      ? setIsOpenModal(true)
      : showErrorNotify(emailRes.msg);
  }, [emailRes, showErrorNotify]);

  return (
    <Container>
      <Layout>
        <Logo>
          <img src="/assets/logo.png" alt="로고" />
        </Logo>
        <Head>회원가입</Head>
        <Form onSubmit={handleSubmit(onSubmit)} action="">
          <Input
            type="email"
            title="이메일"
            placeholder="이메일"
            autoFocus
            {...register("email", {
              required: ERROR_MSG.required,
              pattern: {
                value: EMAIL_REGEX,
                message: ERROR_MSG.invalidEmail,
              },
            })}
          />
          {errors.email && <ErrorMsg>{errors.email.message}</ErrorMsg>}
          <Input
            type="password"
            title="비밀번호"
            placeholder="비밀번호"
            {...register("pw", {
              required: ERROR_MSG.required,
              pattern: {
                value: PW_REGEX,
                message: ERROR_MSG.invalidPw,
              },
              onBlur: () => trigger("confirmPw"),
            })}
          />
          {errors.pw && <ErrorMsg>{errors.pw.message}</ErrorMsg>}
          <Input
            type="password"
            title="비밀번호 확인"
            placeholder="비밀번호 확인"
            {...register("confirmPw", {
              required: ERROR_MSG.required,
              validate: {
                confirmPw: (v) =>
                  v === getValues("pw") || ERROR_MSG.invalidConfirmPw,
              },
            })}
          />
          {errors.confirmPw && <ErrorMsg>{errors.confirmPw.message}</ErrorMsg>}
          <Input
            submit
            type="submit"
            title="회원가입"
            value="회원가입"
            disabled={isSubmitting}
          />
        </Form>
        <Link to="/login">
          <ToLogin>로그인 하러가기</ToLogin>
        </Link>
        {isOpenModal && (
          <SuccessSendMail
            email={getValues("email")}
            password={getValues("pw")}
            closeHandler={closeHandler}
          />
        )}
      </Layout>
    </Container>
  );
};

export default Signup;
