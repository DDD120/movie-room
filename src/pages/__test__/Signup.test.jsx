import { fireEvent, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { renderWithProviders } from "lib/test";
import Signup, { ERROR_MSG } from "pages/Signup";

describe("signup page", () => {
  const setup = () => {
    return renderWithProviders(<Signup />);
  };

  it("renders correctly", () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it("should render 'required' error message", async () => {
    setup();

    const button = screen.getByTitle("회원가입");
    userEvent.click(button);

    const required = await screen.findAllByText(ERROR_MSG.required);
    expect(required).toHaveLength(3);
  });

  it("should render 'invalidEmail' error message", async () => {
    setup();

    const email = screen.getByPlaceholderText("이메일");
    fireEvent.change(email, { target: { value: "1" } });

    const invalidEmail = await screen.findByText(ERROR_MSG.invalidEmail);
    expect(invalidEmail).toBeInTheDocument();
  });

  it("should render 'invalidPw' error message", async () => {
    setup();

    const password = screen.getByPlaceholderText("비밀번호");
    fireEvent.change(password, { target: { value: "1" } });

    const invalidPw = await screen.findByText(ERROR_MSG.invalidPw);
    expect(invalidPw).toBeInTheDocument();
  });

  it("should render 'invalidConfirmPw' error message", async () => {
    setup();

    const password = screen.getByPlaceholderText("비밀번호");
    const confirmPassword = screen.getByPlaceholderText("비밀번호 확인");
    fireEvent.change(password, { target: { value: "1" } });
    fireEvent.change(confirmPassword, { target: { value: "2" } });

    const invalidConfirmPw = await screen.findByText(
      ERROR_MSG.invalidConfirmPw
    );
    expect(invalidConfirmPw).toBeInTheDocument();
  });
});
