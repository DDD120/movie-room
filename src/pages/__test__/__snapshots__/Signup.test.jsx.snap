// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signup page renders correctly 1`] = `
<div>
  <div
    class="css-1ussbir"
  >
    <div
      class="css-jxj5pd"
    >
      <div
        class="css-o5wame"
      >
        <img
          alt="로고"
          src="/assets/logo.png"
        />
      </div>
      <h1
        class="css-wk5jya"
      >
        회원가입
      </h1>
      <form
        class="css-7odvop"
      >
        <input
          class="css-1a6crj1"
          name="email"
          placeholder="이메일"
          title="이메일"
          type="email"
        />
        <input
          class="css-1a6crj1"
          name="pw"
          placeholder="비밀번호"
          title="비밀번호"
          type="password"
        />
        <input
          class="css-1a6crj1"
          name="confirmPw"
          placeholder="비밀번호 확인"
          title="비밀번호 확인"
          type="password"
        />
        <input
          class="css-ol4lpg"
          title="회원가입"
          type="submit"
          value="회원가입"
        />
      </form>
      <a
        href="/login"
      >
        <span
          class="css-hfj296"
        >
          로그인 하러가기
        </span>
      </a>
    </div>
  </div>
</div>
`;
